contracts/SlashIndicator.sol

```diff
@@ -220,29 +242,74 @@ contract SlashIndicator is ISlashIndicator,System,IParamSubscriber, IApplication
+    require(_verifyBLSSignature(_evidence.voteA, _evidence.voteAddr) &&
+      _verifyBLSSignature(_evidence.voteB, _evidence.voteAddr), "verify signature failed");
 
     // reward sender and felony validator if validator found
     (address[] memory vals, bytes[] memory voteAddrs) = IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).getLivingValidators();
     for (uint i; i < voteAddrs.length; ++i) {
       if (BytesLib.equal(voteAddrs[i],  _evidence.voteAddr)) {
         uint256 amount = (address(SYSTEM_REWARD_ADDR).balance * finalitySlashRewardRatio) / 100;
         ISystemReward(SYSTEM_REWARD_ADDR).claimRewards(msg.sender, amount);
-        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).felony( vals[i]);
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).felony(vals[i]);
         break;
       }
     }
 
-    // send slash msg to bc
+    uint256 slashHeight = _evidence.voteA.srcNum;
+    if (_evidence.voteB.srcNum < slashHeight) {
+      slashHeight = _evidence.voteB.srcNum;
+    }
     bytes32 voteAddrSlice = BytesLib.toBytes32(_evidence.voteAddr,0);
-    try ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(SLASH_CHANNELID, encodeVoteSlashPackage(_evidence.voteAddr), 0) {
-      emit maliciousVoteSlashed(voteAddrSlice);
-    } catch (bytes memory reason) {
-      emit failedMaliciousVoteSlash(voteAddrSlice, reason);
+    if (IStakeHub(STAKE_HUB_ADDR).getOperatorAddressByVoteAddress(_evidence.voteAddr) != address(0)) {
+      try IStakeHub(STAKE_HUB_ADDR).maliciousVoteSlash(_evidence.voteAddr, slashHeight) {
+        emit maliciousVoteSlashed(voteAddrSlice);
+      } catch (bytes memory reason) {
+        emit failedMaliciousVoteSlash(voteAddrSlice, reason);
+      }
+    } else {
+      // send slash msg to bc if the validator not migrated
+      try ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(SLASH_CHANNELID, encodeVoteSlashPackage(_evidence.voteAddr), 0) {
+        emit maliciousVoteSlashed(voteAddrSlice);
+      } catch (bytes memory reason) {
+        emit failedMaliciousVoteSlash(voteAddrSlice, reason);
+      }
     }
   }
 
+  function submitDoubleSignEvidence(bytes memory header1, bytes memory header2) public onlyInit {
+    require(header1.length != 0 && header2.length != 0, "empty header");
+
+    bytes[] memory elements = new bytes[](3);
+    elements[0] = bscChainID.encodeUint();
+    elements[1] = header1.encodeBytes();
+    elements[2] = header2.encodeBytes();
+
+    // call precompile contract to verify evidence
+    bytes memory input = elements.encodeList();
+    bytes memory output = new bytes(52);
+    assembly {
+      let len := mload(input)
+      if iszero(staticcall(not(0), 0x68, add(input, 0x20), len, add(output, 0x20), 0x34)) {
+        revert(0, 0)
+      }
+    }
+
+    address signer;
+    uint256 height;
+    uint256 evidenceTime;
+    assembly {
+      signer := mload(add(output, 0x14))
+      height := mload(add(output, 0x34))
+      evidenceTime := mload(add(output, 0x54))
```

Comment:
ensure the "evidenceTime" happens recently

Reply:
Done